source/SH3/system/glprogram.cpp

```diff
@@ -0,0 +1,118 @@
+/** @file
+ *
+ *  Implementation of glprogram.hpp
+ *
+ *  @copyright 2016  Palm Studios
+ *
+ *  @date
+ *
+ *  @author Jesse Buhagiar
+ */
+#include <SH3/system/glprogram.hpp>
+#include <SH3/system/log.hpp>
+
+#include <fstream>
+#include <vector>
+
+using namespace sh3_graphics;
+
+GLint sh3_glprogram::Load(const std::string& name)
+{
+    GLint status = 0;
+
+    // Generate our vertex and fragment shader
+    vertShader = (GLuint)Compile(GL_VERTEX_SHADER);
+    fragShader = (GLuint)Compile(GL_FRAGMENT_SHADER);
+
+    glAttachShader(programID, vertShader);
+    glAttachShader(programID, fragShader);
+    glLinkProgram(programID);
+
+    // Make sure our program was linked without any errors
+    glGetProgramiv(programID, GL_LINK_STATUS, (int*)&status);
+
+    if(status == GL_FALSE)
+    {
+        GLint logLength = 0;
+        std::vector<GLchar> errorLog;   // Error log
+        glGetShaderiv(programID, GL_INFO_LOG_LENGTH, &logLength);
+
+        errorLog.resize(logLength);
+        glGetProgramInfoLog(programID, logLength, &logLength, &errorLog[0]);
+        glDeleteProgram(vertShader);
+        glDeleteShader(fragShader);
+
+        die(&errorLog[0]);
+    }
+
+    glDetachShader(programID, vertShader);
+    glDetachShader(programID, fragShader);
+}
+
+void sh3_glprogram::Bind()
+{
+
+}
+
+void sh3_glprogram::Unbind()
+{
+
+}
+
+GLuint sh3_glprogram::Compile(std::uint32_t type)
+{
+    GLint   status; // Used to check compilation status
+    GLuint  id;     // ID of this shader
+
+    std::ifstream       source;     // Handle to the shader program source
+    std::string         fname;      // Name of the file we want to open
+    std::vector<GLchar> errorLog;   // Error log
+
+    if(type == GL_VERTEX_SHADER)
```

Comment:
There's plenty of code duplication in these branches.
For each shader type, only fname seems to be the difference.